File: embedded/src/hardware/led.go
package hardware

import (
	"alarm_project/embedded/src/config"
	"alarm_project/embedded/src/log"
	"machine"
)

type LED struct {
	pin       machine.Pin
	pwm       machine.PWM
	pwmCh	  uint8
}

func NewLED() *LED {
	led := &LED{
		pin: config.LED_PIN,
		pwm: machine.PWM4,
	}
	led.pwm.Configure(machine.PWMConfig{
		Period: 1e6, // 1MHz frequency
	})
	ch, err := led.pwm.Channel(led.pin)
	if err != nil {
		log.Error("Failed to configure PWM channel for LED: %v", err)
		panic(err)
	}
	led.pwmCh = ch
	led.SetIntensity(0)
	return led
}

// intensity: 0-100 where 0 is off and 100 is full intensity
func (l *LED) SetIntensity(intensity int) {
	// Ensure intensity is within bounds
	if intensity < 0 {
		intensity = 0
	} else if intensity > 100 {
		intensity = 100
	}
	
	// Convert 0-100 range to 0-pwm.Top for PWM
	pwmValue := uint32((float64(intensity) * l.pwm.Top()) / 100.0)
	l.pwm.Set(l.pwmCh, pwmValue)
}

File: embedded/src/hardware/led.py
from machine import Pin, PWM
from config.constants import MAX_DUTY, PWM_FREQ

class LED:
    def __init__(self, pin_number: int):
        self._pwm = PWM(Pin(pin_number))
        self._pwm.freq(PWM_FREQ)
        self._pwm.duty_u16(0)
        self._current_intensity = 0
        
    def set_intensity(self, intensity: int):
        """Set LED intensity from 0-100"""
        self._current_intensity = max(0, min(100, intensity))
        duty = int((self._current_intensity / 100) * MAX_DUTY)
        self._pwm.duty_u16(duty)
    
    def get_intensity(self) -> int:
        """Get current intensity 0-100"""
        return self._current_intensity
    
    def off(self):
        """Turn off the LED"""
        self.set_intensity(0)
    
    def cleanup(self):
        """Clean up resources"""
        self._pwm.deinit()
